### ISSUE TITLE
Test passes but log shows a failed assert

### ISSUE THREAD

**Issue by JessefSpecialisterren** [cypress-io/cypress#9155]
### Current behavior
Under certain circumstances, Cypress shows a test as passed even though the log contains a failed assert:
![image](https://user-images.githubusercontent.com/36038882/98674361-fdf94c80-2358-11eb-905d-0945785b9fad.png)

### Desired behavior
In this particular case, the failed assert should not be shown. It was not executed in the invocation of the `should` callback that made the test succeed, because the third li element had been removed by that time.

### Test code to reproduce
https://github.com/JessefSpecialisterren/cypress-test-tiny/tree/assert-fails-test-succeeds

### Versions
Cypress: 5.6.0
Browser: Chrome 86
OS: Windows 10

**Comment by jennifer-shehane** [comment]
Thanks for the simple example. I can see that the async nature of the app code will cause this situation where Cypress passes the test, and then updates the last assertion to fail.

![](http://g.recordit.co/DjpBrT6vmh.gif)

### Reproducible example

`index.html`

```html
<body>
  <ul>
    <li>foo</li>
    <li>foo</li>
    <li>bar</li>
  </ul>
  <script>
    setTimeout(() => {
      document
        .querySelector('li:last-child')
        .remove()
    }, 1000)
  </script>
</body>
```

`spec.js

```js
it('all li elements have text "foo"', () => {
  cy.visit('index.html')
  cy.get('li').should(($lis) => {
    $lis.each((_, element) => {
      expect(element.textContent).to.equal('foo')
    })
  })
})
```

### Workaround

The ideal way to write this test would be to assert first that the removed element doesn't exist first:

```js
it('all li elements have text "foo"', () => {
  cy.visit('index.html')
  cy.get('li').should('have.length', 2).and(($lis) => {
    $lis.each((_, element) => {
      expect(element.textContent).to.equal('foo')
    })
  })
})
```

**Comment by bahmutov** [comment]
Yeah, this is unexpected. Once the test completes, the assertions should not change their state. In this particular case I would make sure the expected number of assertions has passed as described in https://www.cypress.io/blog/2020/01/16/when-can-the-test-stop/

**Comment by JessefSpecialisterren** [comment]
Note that the real-world testcase that lead to this report types text into a filter field and checks if the filtered list only has elements containing the filter text. The number of results (and hence the number of assertions) is not known beforehand, so the proposed workarounds won't work

**Comment by bahmutov** [comment]
Sure they will. It is just that the test has no idea when the application updates itself and when the test should finish. What if the app updates itself after 10 seconds? Or 15 seconds? So you have to observe something for the test to "know" when it can check the filtered results. Worst comes to worst, you can always add `cy.wait(20000)` before asserting the filtered results. I would do things a little differently - I would use cy.route2 to spy on the network request, once the network call returns I could look at the number of returns results and then check the DOM. Now you know how many results to expect and can verify it before checking each shown item.

**Comment by JessefSpecialisterren** [comment]
I settled for this workaround:
```js
it('all li elements have text "foo"', () => {
  cy.visit('index.html')
  cy.get('li').should(($lis) => {
    let nonFoo = []
    $lis.each((_, element) => {
      if (element.textContent !== 'foo') nonFoo.push(element)
    })
    expect(nonFoo, 'non-foo lis').to.be.empty
  })
})
```
Though I agree with @bahmutov that in general, it would be better to wait until the DOM is stable. My current project is effectively black box testing, however, so I prefer a solution that only looks at the html

**Comment by JessefSpecialisterren** [comment]
Verified the behavior is still occuring and updated the test code

**Comment by nagash77** [comment]
@JessefSpecialisterren thank you for confirming this issue is still happening.  I will update the ticket.

**Comment by chrisbreiding** [comment]
I replicated this issue in the latest version of Cypress and will forward this ticket to the appropriate team. They will evaluate the priority of this ticket and consider their capacity to pick it up. Please note that this does not guarantee that this issue will be resolved.
